File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/consumer/acknowledge/type/AcknowledgeSpec.java
package io.micronaut.rabbitmq.docs.consumer.acknowledge.type;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

@MicronautTest
@Property(name = "spec.name", value = "AcknowledgeSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class AcknowledgeSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testAckingWithAcknowledgement(ProductClient productClient, ProductListener productListener) {

// tag::producer[]
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
productClient.send("message body".getBytes());
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                productListener.messageCount.get() == 5 // the first message is rejected and re-queued
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/consumer/acknowledge/type/AcknowledgeSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="AcknowledgeSpec")
class AcknowledgeSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_acking_with_acknowledgement(self) -> None:
# tag::producer[]
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
        self.product_client.send(b"message body")
# end::producer[]

        # the first message is rejected and re-queued
        Await.until(lambda: self.product_listener.message_count == 5)
